---
title: "MMPChien Lab scRNAseq pipeline code | Interaction Analysis"
author: "Myrthe Smit"
output:
  html_document:
    highlight: pygments
    df_print: paged
    toc: yes
    toc_float: true
    number_sections: true
---

```{r setup, include=FALSE}
# Erase everything in the global environment
# It is a good habit to start with a clean slate
rm(list = ls())

# Set the knitr options
knitr::opts_chunk$set(
  echo = TRUE,
  error = TRUE,
  message = FALSE,
  warning = FALSE,
  collapse = TRUE,
  results = "hold",
  fig.keep = "all",
  fig.show = "hold",
  comment = NA
)
```

<span style="color:#cd6155;">**Libraries**</span> <br>
For this analysis we will use the packages:<br>
- [clusterProfiler](http://yulab-smu.top/clusterProfiler-book/index.html)
- [msigdbr](https://cran.r-project.org/web/packages/msigdbr/vignettes/msigdbr-intro.html) to retrieve the reference database <br>
- [fgsea](https://github.com/ctlab/fgsea) to perform GSEA

```{r libraries}
library(biomaRt)
library(clusterProfiler)
library(ComplexHeatmap)
library(enrichplot)
library(fgsea)
library(future)
library(GSVA)
library(msigdbr)
library(patchwork)
library(RColorBrewer)
library(Seurat)
library(tidyverse)
library(viridis)

# Set the amount of cores to use
plan("multiprocess", workers = 5)
# Specify the amount of memory to use (in mbs)
MEMORY <- 700
options(future.globals.maxSize = MEMORY * (1024^2))

rm(MEMORY)
```

# The Experimental Information
<span style="color:#cd6155;">**scRNAseq_analysis_code version**</span> v5.1

<span style="color:#cd6155;">**Experimental Code**</span> <br>
MS_SpatialTranscriptomics

<span style="color:#cd6155;">**Description of the wet lab experiment and the data**</span> <br>
In this experiment we seeded a patch of MCF10A cells and let them grow for 6 days. Then we phototagged cells with two different phototagging dyes to get 3 regions: the outside ring of the clump, the middle ring of the clump and the center. In one set-up, we made equally-sized rings and in the other set-up we made rings with width 250 um. 

<span style="color:#cd6155;">**Goal of this analysis **</span> <br>
Test if the cells at the edge have a higher level of EMT than cells in the center.

<span style="color:#cd6155;">**Sample description**</span> <br>
```{r experiments, echo=FALSE}
# Read in the Excel file with the sample information
sampleinformation <- readxl::read_excel("2_scData/SampleOverview.xlsx")

# Format the table
formattable::formattable(sampleinformation, align = c("l"))
rm(sampleinformation)
```

# The Basis | Directories and Data
## Directories and project name
**Directories** <br>
```{r}
dataDirectory <- "2_scData/"
genesDirectory <- "3_GenesofInterest/"
interactionDirectory <- "6_Interactions/"
imageDirectory <- "_Figures"

# These names will be used in file saving
# PROJECTNAME <- "MS_SpatialTranscriptomics"
PROJECTNAME <- "MS011"
ANALYSISNAME <- "edge"
```

# Load data
```{r}
corrected_object <- readRDS(file = paste0(dataDirectory, PROJECTNAME, "_", ANALYSISNAME, "_CorrectedObject.rds"))

seurat_data <- list()

for (i in names(corrected_object)) {
  seurat_data[[i]] <- GetAssayData(corrected_object[[i]], assay = "RNA", slot = "data")
}

seurat_matrix <- as.matrix(seurat_data$MS011)
```

# Prepare CellphoneDB analysis 
We need to prepare a gene expression matrix and a metadata file containing the cells and their populations. This needs to be formatted in a specific way so that the package understands it. 
```{r}
# write.table(
#   seurat_matrix,
#   paste0(interactionDirectory, PROJECTNAME, "_", ANALYSISNAME, "_cellphonedb_count.txt"),
#   sep = "\t",
#   quote = F,
#   row.names = T
# )

cellnames <- as.data.frame(colnames(seurat_matrix))
treatments <- as.data.frame(corrected_object[["MS011"]]@meta.data[["treatment"]])

metadata <- cbind(cellnames, treatments)
colnames(metadata) <- c("Cell", "cell_type")

# write.table(
#   metadata,
#   paste0(interactionDirectory, PROJECTNAME, "_", ANALYSISNAME, "_cellphonedb_meta.txt"),
#   sep = "\t",
#   quote = F,
#   row.names = F
# )
```

Now the actual CellphoneDB analysis can be ran on Ubuntu (on Myrthe's computer).

# Visualization
Modified version of the [CellphoneDB](https://www.cellphonedb.org) dotplot function so that we get nice plots
```{r}
dot_plot <- function(selected_rows = NULL,
                    selected_columns = NULL,
                    filename = "plot.pdf",
                    width = 8,
                    height = 10,
                    means_path = "./means.txt",
                    pvalues_path = "./pvalues.txt",
                    means_separator = "\t",
                    pvalues_separator = "\t",
                    output_extension = ".pdf") {
  all_pval <- read.table(pvalues_path, header = T, stringsAsFactors = F, sep = means_separator, comment.char = "", check.names = F)
  all_means <- read.table(means_path, header = T, stringsAsFactors = F, sep = pvalues_separator, comment.char = "", check.names = F)

  intr_pairs <- all_pval$interacting_pair
  all_pval <- all_pval[, -c(1:11)]
  all_means <- all_means[, -c(1:11)]

  if (is.null(selected_rows)) {
    selected_rows <- intr_pairs
  }

  if (is.null(selected_columns)) {
    selected_columns <- colnames(all_pval)
  }

  sel_pval <- all_pval[match(selected_rows, intr_pairs), selected_columns]
  sel_means <- all_means[match(selected_rows, intr_pairs), selected_columns]

  df_names <- expand.grid(selected_rows, selected_columns)
  pval <- unlist(sel_pval)
  pval[pval == 0] <- 0.0009
  plot.data <- cbind(df_names, pval)
  pr <- unlist(as.data.frame(sel_means))
  pr[pr == 0] <- 1
  plot.data <- cbind(plot.data, log2(pr))
  colnames(plot.data) <- c("pair", "clusters", "pvalue", "mean")

  # my_palette <- colorRampPalette(c("black", "blue", "yellow", "red"), alpha=TRUE)(n=399)
  my_palette <- viridis_pal()(399)
  # my_palette <- brewer.pal(n = 11, name = "PuBu")

  ggplot(plot.data, aes(x = clusters, y = pair)) +
    geom_point(aes(size = -log10(pvalue), color = mean)) +
    scale_color_gradientn("Log2 mean (Molecule 1, Molecule 2)", colors = my_palette) +
    theme_bw() +
    theme(
      panel.grid.minor = element_blank(),
      panel.grid.major = element_blank(),
      axis.text = element_text(size = 8, colour = "black"),
      axis.text.x = element_text(angle = 90, hjust = 1),
      axis.text.y = element_text(size = 8, colour = "black"),
      axis.title = element_blank(),
      panel.border = element_rect(size = 0.5, linetype = "solid", colour = "black")
    )

  # if (output_extension == ".pdf") {
  #   ggsave(filename, width = width, height = height, device = cairo_pdf, limitsize = F)
  # } else {
  #   ggsave(filename, width = width, height = height, limitsize = F)
  # }
}

```

Make the plot
```{r}
# select which interaction pairs you want to display
selected_genes <- read.csv("./6_Interactions/MS011_edge_out/figure3_genes.csv", header = F)
selected_genes <- unlist(selected_genes, use.names = F)

# select which cell-cell interactions you want to display
selected_cells <- read.csv("./6_Interactions/MS011_edge_out/selected_cells.csv", header = F)
selected_cells <- unlist(selected_cells, use.names = F)

fig_width <- 4.7 + (length(selected_cells) * 0.25)
fig_height <- 0.7 + (length(selected_genes) * 0.25)

pvalues <- "./6_Interactions/MS011_edge_out/pvalues.txt"
means <- "./6_Interactions/MS011_edge_out/means.txt"

dot_plot(
  means_path = means,
  pvalues_path = pvalues,
  filename = "edge_dotplot.pdf",
  width = fig_width,
  height = fig_height,
  selected_rows = selected_genes,
  selected_columns = selected_cells
)
```

<details>
  <summary>Session information</summary>
```{r}
devtools::session_info()
```
</details> 