---
title: "MPChien Lab scRNAseq pipeline code | Seurat object creation"
author: "Kate J. Feller; modified by Myrthe Smit"
output:
  html_document:
    highlight: pygments
    df_print: paged
    toc: yes
    toc_float: true
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  error = TRUE,
  message = FALSE,
  warning = FALSE,
  results = "hold",
  fig.keep = "all",
  fig.show = "hold",
  comment = NA
)
```

<span style="color:#cd6155;">**Libraries**</span> <br>
The MP Chien scRNAseq pipeline is build around the [Seurat package](https://satijalab.org/) standard scRNAseq analysis pipeline.
```{r libraries}
rm(list = ls())

# Seurat library
library(Seurat)

# General libraries
library(data.table)
library(formattable)
library(future)
library(readxl)
library(stringr)
library(tidyverse)

# The amount of CPU cores to use
plan("multiprocess", workers = 5)
# Specify the amount of memory (RAM) to use in mb
MEMORY <- 700
options(future.globals.maxSize = MEMORY * (1024^2))

rm(MEMORY)
```

# The Experimental Information
<span style="color:#cd6155;">**scRNAseq_analysis_code version**</span> v5.1

<span style="color:#cd6155;">**Experimental Code**</span> <br>
MS_SpatialTranscriptomics

<span style="color:#cd6155;">**Description of the wet lab experiment and the data**</span> <br>
In this experiment we seeded a patch of MCF10A cells and let them grow for 6 days. Then we phototagged cells with two different phototagging dyes to get 3 regions: the outside ring of the clump, the middle ring of the clump and the center. In one set-up, we made equally-sized rings and in the other set-up we made rings with width 250 um. 

<span style="color:#cd6155;">**Goal of this analysis **</span> <br>
Test if the cells at the edge have a higher level of EMT than cells in the center. In this code we will generate a Seurat object.

<span style="color:#cd6155;">**Sample description**</span> <br>
```{r experiments, echo=FALSE}
# Read in the Excel file with the sample information
sampleinformation <- read_excel("2_scData/SampleOverview.xlsx")

# Format the table
formattable::formattable(sampleinformation, align = c("l"))
rm(sampleinformation)
```

# The Basis | Directories and Data
## Directories and project name
Specify where to read and save the data and generated outputs to.
```{r}
# Where to read the scRNAseq data from
dataDirectory <- "2_scData/CountTables/"
# Where to read the sample annotation from
sampleDirectory <- "2_scData/SampleNames/"

# The project name will be used in file saving
PROJECTNAME <- "MS_SpatialTranscriptomics"
```

## Importing the data
For this analysis we will use `TranscriptCounts.tsv` files received from [Single Cell Discoveries, Utrecht](https://www.scdiscoveries.com/) and these contain the Poisson corrected UMI counts.

```{r}
# Identify all "TranscriptCounts.tsv" files
rawCount_files <- list.files(path = dataDirectory, pattern = "*.TranscriptCounts.tsv", full.names = TRUE)
print(paste("There are", length(rawCount_files), "data files", sep = " "))

rawCounts_Data <- list()

for (file in rawCount_files) {
  fileID <- str_extract(str_extract(file, "[ERA-MC-s](\\d+)"), "\\d+") # select the plate ID
  print(fileID)
  rawCounts_Data[[fileID]] <- read_tsv(file = file)
}

rm(rawCount_files, file)
```

Next we will import the sample annotations as generated by the Python `SampleAnnotation_v1.1.py` file. The output of the Python file is standardized in the format `xxx_TG_sample`, where:

- xxx &#8594; 3 digits of the SORTseq plate ID
- TG &#8594; the phototagging information where `NT = not tagged` and `PT = phototagged`
- sample &#8594; the user specified sample name
- in this file wells that contained no cells are annotated as `empty`

```{r}
# Identify all files containing the sample names
sampleName_files <- list.files(path = sampleDirectory, pattern = "*.xlsx", full.names = TRUE)
print(paste("There are", length(sampleName_files), "sample name files", sep = " "))

sampleNames <- list()

for (file in sampleName_files) {
  fileID <- str_extract(file, "[0-9]{3}") # select only the numbers in the name = plate ID
  sampleNames[[fileID]] <- read_excel(file, sheet = "Sheet1", col_names = FALSE)

  rm(fileID)
}

rm(sampleName_files, file)
```

# Preparing the data
Overview of the steps:

1. merge the sample names and the transcriptcounts data
2. make the column names == sample names
3. merge individual batches/plates into a single dataframe
4. replace any NA's in data by zeros
5. make the row names == HGNC gene symbols
6. filter out the SpikeIn genes (ERCC genes) and empty columns/wells

```{r}
nameddata <- list()

for (id in names(rawCounts_Data)) {
  data <- rawCounts_Data[[id]]

  # Replace the COLUMN NAMES
  colNames <- t(data.frame(make.unique(sapply(sampleNames[[id]], as.character), sep = ".")))
  countMatrix <- data.frame(data)
  colnames(countMatrix)[2:ncol(countMatrix)] <- colNames

  nameddata[[id]] <- countMatrix

  rm(data, rowNames, colNames, countMatrix)
}
rm(id, rawCounts_Data, sampleNames)
```

```{r}
# Merge into one large dataframe
alldata <- nameddata %>%
  reduce(full_join, by = "GENEID")

# 1. REPLACE NA VALUES
alldata[is.na(alldata)] <- 0

# 2. ROW NAMES
rowNames <- alldata[1]
rowNames$GENEID <- sub("\\__.*", "", rowNames$GENEID)
rowNames <- t(rowNames)
## Make the row names unique as sometimes this is not always the case
rowNames <- t(data.frame(make.unique(rowNames, sep = "..")))

# 3. CREATE THE CLEAN DATA
## Make sure the data is numeric
countMatrix <- alldata[sapply(alldata, is.numeric)]
## Add in the new and correct row and column names
countMatrix <- data.frame(countMatrix, row.names = rowNames)

# 4. REMOVE SPIKEINS AND EMPTY COLUMNS
## Remove the Spike In genes
ERCCcounts <- row.names(countMatrix[grep("^ERCC-", row.names(countMatrix)), ])
print(paste("The amount of ERCC rows is", length(ERCCcounts), sep = " "))
countMatrix <- countMatrix[!rownames(countMatrix) %in% ERCCcounts, ]

## Remove the empty columns (= empty scPlate wells)
## This code assumes that empty wells contain the word "empty" in the name
EMPTY_COLS <- grep("empty", colnames(countMatrix))
print(paste("The amount of empty columns is", length(EMPTY_COLS), sep = " "))
if (length(EMPTY_COLS) > 0) {
  cleanedcountMatrix <- countMatrix[, -EMPTY_COLS]
} else {
  cleanedcountMatrix <- countMatrix
}
print(paste("This dataset contains", ncol(cleanedcountMatrix), "cells", sep = " "))

# Remove the unnecessary variables from the environment
rm(rowNames, countMatrix, ERCCcounts, EMPTY_COLS)
```

# Creating a Seurat object with informative meta_data
## Seurat object
```{r}
beforeQC_object <- CreateSeuratObject(
  counts = cleanedcountMatrix,
  min.cells = 3,
  min.features = 0,
  project = PROJECTNAME
)

head(beforeQC_object@meta.data)
```

## Metadata
In this dataset the names, which were imported from the sample name excel files, are formatted as `scplate_phototagging_sampleID`(see the first column in the output of the previous section)

```{r}
## Positions in the sample annotation string
scPLATE_POSITION <- c(2, 4)
TREATMENT_POSITION <- c(6, 7)
SAMPLEID_POSITION <- c(9, 11)

## METADATA
# Extract the sample information
rawMetaData <- beforeQC_object@assays$RNA@counts@Dimnames[[2]]

# Create a column for the scPlate number (which also equals the batch)
beforeQC_object@meta.data$batch <- as.integer(
  substr(rawMetaData,
    start = scPLATE_POSITION[[1]],
    stop = scPLATE_POSITION[[2]]
  )
)

# Create a column for the sampleID
beforeQC_object@meta.data$sample <- substr(
  rawMetaData,
  start = SAMPLEID_POSITION[[1]],
  stop = SAMPLEID_POSITION[[2]]
)

# Create a column for the treatment or other difference between samples
beforeQC_object@meta.data$treatment <- substr(
  rawMetaData,
  start = TREATMENT_POSITION[[1]],
  stop = TREATMENT_POSITION[[2]]
)

## NOTE: for this experiment I need to rewrite the treatment ID so it made sense for analysis
beforeQC_object@meta.data$treatment <- ifelse(grepl("cen", beforeQC_object@meta.data$sample), "inner", ifelse(grepl("mid", beforeQC_object@meta.data$sample), "middle", ifelse(grepl("out", beforeQC_object@meta.data$sample), "outer", "unknown")))

rm(rawMetaData, scPLATE_POSITION, TREATMENT_POSITION, SAMPLEID_POSITION)
```

```{r}
table(beforeQC_object@meta.data$sample)
```

# Save the Seurat object
```{r class.source = "fold-show"}
# saveRDS(beforeQC_object, file = paste0("2_scData/", PROJECTNAME, "_BeforeQC.rds"))
```

<details>
  <summary>Session information</summary>
```{r}
devtools::session_info()
```
</details> 

